crates/autopilot/src/database/onchain_order_events/ethflow_events.rs

```diff
@@ -0,0 +1,208 @@
+                let meta = match meta {
+                    Some(meta) => meta,
+                    None => return Some(Err(anyhow!("event without metadata"))),
+                };
+                let ContractEvent::OrderPlacement(event) = data;
+                let quote_and_valid_to_touple = convert_to_quote_id_and_user_valid_to(event);
+                match quote_and_valid_to_touple {
+                    Ok(touple) => Some(Ok((
+                        meta_to_event_index(meta),
+                        CustomParsedOnchaninData {
+                            quote_id: touple.0,
+                            additional_data: Some(touple.1),
+                        },
+                    ))),
+                    Err(err) => {
+                        tracing::debug!(
+                            "Error while converting quote id and user valid to: {:?}",
+                            err
+                        );
+                        None
+                    }
+                }
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+
+    async fn append_custom_order_info_to_db(
+        &self,
+        ex: &mut PgTransaction<'a>,
+        custom_onchain_data: Vec<EthOrderPlacement>,
+    ) -> Result<()> {
+        database::ethflow_orders::append(ex, custom_onchain_data.as_slice())
+            .await
+            .context("append_ethflow_orders failed")
+    }
+
+    fn customized_event_data_for_event_index(
+        &self,
+        event_index: &EventIndex,
+        order: &Order,
+        hashmap: &HashMap<EventIndex, u32>,
+        _onchain_order_placement: &OnchainOrderPlacement,
+    ) -> EthOrderPlacement {
+        EthOrderPlacement {
+            uid: order.uid,
+            // unwrap is allowed, as any missing event_index would have been filtered beforehand
+            // by the implementation of the function parse_custom_event_data
+            valid_to: *hashmap.get(event_index).unwrap() as i64,
+        }
+    }
+}
+
+fn convert_to_quote_id_and_user_valid_to(
+    order_placement: &ContractOrderPlacement,
+) -> Result<(i64, u32)> {
+    let user_valid_to_bytes = TryInto::<[u8; 4]>::try_into(
+        order_placement
+            .data
+            .0
+            .iter()
+            .take(4)
+            .copied()
+            .collect::<Vec<u8>>(),
+    )
+    .map_err(|err| anyhow!("Error while decoding user_valid_to data: {:?}", err))?;
+    println!("{:?}", user_valid_to_bytes);
+    let user_valid_to = u32::from_be_bytes(user_valid_to_bytes);
+    let quote_id_bytes = TryInto::<[u8; 8]>::try_into(
+        order_placement
+            .data
+            .0
+            .iter()
+            .skip(4)
+            .take(8)
+            .copied()
+            .collect::<Vec<u8>>(),
+    )
+    .map_err(|err| anyhow!("Error while decoding user_valid_to data: {:?}", err))?;
+
+    let quote_id = i64::from_be_bytes(quote_id_bytes);
+    Ok((quote_id, user_valid_to))
```

Comment:
```suggestion
    let data = order_placement.data.0.as_slice();
    anyhow::ensure!(data.len() == 12, "invalid data length");
    let user_valid_to = u32::from_be_bytes(data[0..4].try_into().unwrap());
    let quote_id = i64::from_be_bytes(data[4..12].try_into().unwrap());
    Ok((quote_id, user_valid_to))
```

The `ensure` in line 2 should maybe assert a different length or `>= 12` instead of `==`. I don't know what data is in there.